### ISSUE TITLE
Observability and concurrency ownership for evaluation

### ISSUE THREAD

**Issue by DEVunderdog** [DEVunderdog/DataForgeML#491]
Part of #486

## Question

Does the new evaluation module emit Pipeline Events, and does it take an `observer=`?

The precedent points both ways and needs resolving deliberately.

- **For:** `EvaluationOrchestrator` is a Phase Orchestrator and every one of them emits Events (CONTEXT.md, Observability). C2ST across 20 columns × 10 repeats is genuinely long-running — the exact "is it hung or working?" case Events exist for. `substep` granularity (ADR-0055) was designed for precisely this shape of inner loop.
- **Against:** ADR-0074 dropped the event stream from `fit_unit` in favour of a returned record, on the reasoning that a caller driving the loop already knows where it is. If evaluation is a single call over a known column set, the caller knows the column count going in and the result record carries the detail.

Sub-questions:

- If Events stay, what is the item and what is the substep? Column as item and repeat as substep is the obvious mapping, but 20 × 10 = 200 substep emits may be noise rather than signal.
- Does an `observer=` on a *standalone* `EvaluationConfig` sit oddly, given every other observer is passed to an orchestrator constructor?
- Is there a `FitSignals`-shaped answer here — the per-run detail the caller cannot see, returned structurally rather than streamed? If so it overlaps with the result record, so this ticket and [the result record ticket] must agree.
- Concurrency: #435's audit called evaluation's own `ThreadPoolExecutor` the last library-owned pool after ADR-0075/0081. The new module must not reinstate one. If the `R` repeats are parallelisable, whose pool runs them?

The last point is not optional — it is a defect the map has already committed to not repeating.

**Comment by DEVunderdog** [comment]
## Answer

**Evaluation emits Pipeline Events, because the library owns the loop. It owns no pool, because the classifier is already threaded. The entry point is a free function taking `observer=`, and it is an umbrella on all three sides — name, config and return.**

### 1. Events stay, and ADR-0074 is not a precedent against them

The ticket framed the precedent as pointing both ways. It does not. ADR-0074's reasoning — *"a caller driving the loop already knows where it is"* — is conditional on **who owns the loop**, not on which phase you are in.

For `fit_unit` the user owns the loop: they call it once per unit and count their own units, so a stream telling them "unit 3 of 7" reports information they already hold. For evaluation the user makes **one call** and the library then iterates columns × frames × repeats internally. The position-in-the-call is not theirs to know. `StructuralProfiler` keeps Events for the identical reason.

So the rule is **observability follows loop ownership**, and it should be written down in exactly those words, because otherwise two idioms sit side by side in the library and the split looks arbitrary.

*Given up:* the new module carries emitter plumbing that `fit_unit` shed, and a reader must learn which idiom applies where.

### 2. The entry point is a free function, not an orchestrator class

[#490](https://github.com/DEVunderdog/DataForgeML/issues/490) deferred this here. The map already ruled the entry point **stateless**, and a class whose only instance state is the observer is constructor ceremony wrapped around a single call.

The `CONTEXT.md` rule reads *"supplied per orchestrator as a live `observer=` constructor argument — deliberately not a field on `PipelineConfig`"*. Read the **reason** rather than the letter: the observer is kept off the config because a live callable would break that config's serializable, setter-only contract (ADR-0044). A keyword argument on a free function satisfies that reason exactly as well as a constructor does.

This answers the ticket's sub-question directly: **`observer=` never goes near `EvaluationConfig`.** It goes on the call. There was never an oddness to resolve.

It also matches where the library has been moving — ADR-0071 collapsed the imputation orchestrator into `decide`/`fit_unit`, ADR-0072 collapsed persistence into `serialize`/`deserialize`/`inspect`.

*Given up:* `StructuralProfiler` stays a class, so the two surviving phases differ in shape; `EvaluationOrchestrator` **disappears from the Public API**, a break beyond the ones the map listed; and `CONTEXT.md`'s Observability section must be amended from *"per orchestrator constructor"* to *"per orchestrator call"*, since **Phase Orchestrator** no longer implies a class.

### 3. Concurrency: nobody runs the repeats, and that is not a compromise

[#435](https://github.com/DEVunderdog/DataForgeML/issues/435)'s audit named evaluation's `ThreadPoolExecutor` the last library-owned pool. The obvious escape — hand the pool to the user — does not work here: the user makes one whole-table call, so owning a pool would mean owning the column loop, which would mean exposing per-column evaluation and unbuilding the entry point #490 just fixed.

**A measured fact resolves it instead.** On this repo's pinned scikit-learn 1.9.0, `HistGradientBoostingClassifier.__init__` has **no `n_jobs` parameter** (verified: the signature is `categorical_features, class_weight, early_stopping, interaction_cst, l2_regularization, learning_rate, loss, max_bins, max_depth, max_features, max_iter, max_leaf_nodes, min_samples_leaf, monotonic_cst, n_iter_no_change, random_state, scoring, tol, validation_fraction, verbose, warm_start`). It parallelises internally over OpenMP threads and takes the whole box on every fit.

So an outer pool fanning columns out would not add parallelism to an idle machine. It would **oversubscribe a machine the inner fits already saturate** — precisely the failure ADR-0081 was written to arithmetic its way out of.

**Decision: no pool, no `max_workers` dial, nowhere in the evaluation module.** The column loop, the `m` loop and the `R` loop are all plain sequential `for` loops. Parallelism is spent entirely inside the classifier, which is where the work is.

*Gained:* ADR-0075 and ADR-0081 need no amendment and no `core_budget` analogue — there is no budget to compute when exactly one estimator is ever fitting. Determinism is free, with no ordering caveat like [#373](https://github.com/DEVunderdog/DataForgeML/issues/373)'s. The library owns zero pools, full stop.

*Given up:* this holds for HGB specifically. A user who **injects** a single-threaded classifier — a plain `DecisionTreeClassifier`, say — gets a sequential run on one core and the library will not offer them a way to widen it. That is ADR-0075's informed-consent shape applied to an injected estimator, and it belongs in the docstring on the `classifier` argument, not in a guard.

### 4. There is no `FitSignals`-shaped answer here

The ticket asked whether there is one, and flagged the overlap with the result record. The overlap is **total**, for a structural reason.

`FitSignals` exists because `fit_unit` returns an **artifact** — the `FittedUnit` — that outlives the call. How-the-fit-went could not be attached to that artifact without becoming a stale claim (ADR-0074), so it needed a separate ephemeral home riding alongside. **Evaluation returns no artifact.** It returns a report, and the report *is* the record of how the run went: `train_accuracy`, `test_accuracy`, `degenerate`, `repeat_z_spread`, `n_test` are already fields on `C2STResult` per [#488](https://github.com/DEVunderdog/DataForgeML/issues/488).

**No signals object.** [#488](https://github.com/DEVunderdog/DataForgeML/issues/488) needed no negotiation with this ticket and is untouched.

*Given up:* run-level facts that are not per-column have no home — wall time, how many columns refused, the BH threshold applied. If wanted they become fields on `C2STReport`, never a second object.

### 5. Granularity: item is the column, Substep is the repeat — and the `Emitter` crosses the generic seam

The loop is column → frame `m` → repeat `R` → one classifier fit (five under `cross_validation`). The `item` is the column.

For the Substep, the frame is the wrong floor: `m` is 1 today, so it would emit exactly one Substep per column — a duplicate of the `item` event — while the `R` fits, the entire long stretch, stay silent. It would report what you can already see and hide what you cannot.

The repeat is the right floor, and it can only be emitted from inside `c2st`, because #490 put the `R` loop there. **That is fine, and the `Emitter` source is why:** its constructor takes `phase` and `stage` as **plain strings**, not `PipelinePhase`, not anything typed. It is already phase-agnostic infrastructure and does not know which phase it serves. #490's rule was that the generic function knows nothing *imputation*-shaped, and an `Emitter` is not imputation-shaped.

Threading it into deep helpers is the established pattern and the literal title of ADR-0055, *"substep granularity and explicit emitter threading"*; today's `_score_mice_cv` already takes one.

**`c2st` gains a trailing `emitter: Emitter | None = None`** and emits one Substep per repeat, riding the existing `index`/`total` fields as `R` and naming the column.

At 20 columns × 10 repeats that is **200 Substep emits, and that is the correct volume.** Substeps log at DEBUG; the observer filters `event_type` in its own code; `CONTEXT.md` is explicit that there are **no library-side on/off knobs** for observability. A repeat is also not the solver-iteration granularity Substep excludes — the library owns and drives every one of these fits, unlike a MICE round inside an estimator.

*Given up:* `_c2st.py` now imports from `observability`, so the generic module is no longer importable in isolation — lifting C2ST into a standalone package would carry the event spine with it.

### 6. Refused columns tick the bar, and say nothing more

`total` on the `item` events is the count of **active columns**, since [#489](https://github.com/DEVunderdog/DataForgeML/issues/489) gives every active column a record and untestable is a result rather than an absence. That forces the rest: a refused column that emitted no `item` would leave the bar short of its own total, reading as a hang at the finish. **So refused columns emit their `item`.**

They emit **nothing else**. A `decision` event carrying the refusal reason was considered and rejected: the reason is already on the record, the record is returned, and duplicating it into the stream creates two renderings of the same outcome vocabulary that can drift. **The event stream is progress only.** Consequence, accepted: on a 190-column table the bar reaches 190/190 in a flash with no on-screen hint that 185 were refused — the user reads that off the report, which is where it lives.

This ticket therefore adds **no new caller** to the Rendering Contract ([#496](https://github.com/DEVunderdog/DataForgeML/issues/496)).

### 7. A free function has nowhere to emit stage boundaries from — so `Emitter` grows two methods

Surfaced while reading the source, not anticipated by the ticket. `Emitter` has only `item` and `substep`. `stage_start` and `stage_end` live on `_ObservabilityMixin`, which is a **class** mixin the profiler inherits. §2 chose a free function, which can inherit nothing, and `CONTEXT.md` requires stage boundaries **always**.

Building a throwaway private object purely to inherit the mixin is the constructor ceremony §2 rejected, wearing a disguise.

**`Emitter` gains `stage_start()` and `stage_end()`, purely additively.** `_ObservabilityMixin` is left exactly as it is, so the profiler does not move; it becomes a class-shaped convenience over an `Emitter` that is now the complete emit surface — usable from a function or a method, which any future free-function phase will need too.

*Given up:* two overlapping ways to emit a stage boundary now exist, and nothing forces the profiler to converge on the newer one. A deliberate duplication left standing rather than paid down.

### 8. `phase="evaluation"`, `stage="c2st"`, and no new `PipelinePhase` member

`PipelinePhase` enumerates the **six sequential pipeline phases** and its job is feeding `resolve_active_columns`. Evaluation is opt-in, out of band, and per #490 resolves its columns against `PipelinePhase.Imputation` — it evaluates *the imputation phase's* column set; it is not a phase of its own. The event's `phase` field is a plain `str`, so nothing requires an enum member.

Naming the stage after the **metric** rather than the module means Energy Distance later lands as a sibling stage under the same phase, with its own `stage_start`/`stage_end` pair, and a multi-metric run reads as several stages without a new field.

*Given up:* `phase` is now sometimes a `PipelinePhase` value and sometimes a bare string with no enum behind it, so the field's domain is looser than it looks — needs a sentence in `CONTEXT.md` so a reader does not read `"evaluation"`'s absence from the enum as an oversight.

### 9. The entry point is an umbrella on all three sides

The name `evaluate_imputation` is an umbrella, which surfaced the real inconsistency in the first draft of this answer: an umbrella **name**, over an umbrella **config** (`EvaluationConfig` with `C2STConfig` nested, exactly so Energy Distance lands as a sibling), returning a **metric-specific record**. Two of three were umbrellas.

The name is right. The alternative — a public `c2st_imputation` today — means one public function per metric, five doors instead of one, contradicting ADR-0050's single-door reading, and adding the umbrella later becomes a breaking rename.

So the **return** becomes an umbrella too:

```python
def evaluate_imputation(
    original_df: pl.DataFrame,
    imputed_frames: Sequence[pl.DataFrame],
    profile: StructuralProfileResult,
    *,
    config: EvaluationConfig,
    metrics: Sequence[EvaluationMetric] | None = None,
    observer: Observer | None = None,
) -> EvaluationReport
```

`EvaluationReport` holds `c2st: C2STReport | None` as its first and today only field. Energy Distance later becomes `report.energy_distance` — purely additive, no type change, nothing existing breaks. This does **not** reopen #488; `C2STReport` is untouched and the envelope sits above it.

*Given up, honestly:* a one-field wrapper dataclass exists today and every read pays an attribute hop. [#447](https://github.com/DEVunderdog/DataForgeML/issues/447) deleted three unread fields rather than keep speculative structure — the difference is that speculative generality builds for a *maybe*, and four further metrics are named, scoped and on this map. The alternative was returning `C2STReport` now and taking a breaking return-type change when the second metric lands.

### 10. `metrics=` selects, and that does not violate accuracy-over-speed

The umbrella running everything raised the obvious objection: what if you want C2ST alone?

**Accuracy-over-speed governs computations that feed a *library* decision.** Profiling computes every indicator always because routing gets better with each one, and switching one off silently degrades the decision. Evaluation metrics are not that shape: they are independent answers to *different questions* handed to a *human*. C2ST answers "is the filled data distinguishable", Per-Column W₁ answers "how far is this column's marginal", Energy Distance answers "one number for the table". Running one does not make another's answer more correct, and the map already describes this surface as *"opt-in, stateless, and re-run repeatedly with different knobs while poking at results"*. Narrowing is **scope selection on an opt-in diagnostic**, not staging a decision to save cost.

**Selection is a `metrics=` argument on the call, defaulting to `None` meaning every metric the library has.** The bare call runs everything, so the umbrella honours its name and nobody needs to know the list exists; one explicit argument narrows it. `EvaluationMetric` is Public API under ADR-0050's input-enum rule, since the user supplies it.

Rejected: a **nullable nested config** (`EvaluationConfig.c2st = None` means off). It makes *absence of configuration* and *refusal of the metric* the same sentinel — the arithmetic-collapse shape #488 rejected one layer down — and puts a control-flow switch inside a declarative, setter-only config.

**The return side follows:** `EvaluationReport.c2st is None` means **not requested**, which is deliberately a different state from *ran and refused* — the latter is #489's six-member outcome vocabulary at column level. Not-run and refused must never collapse into one reading.

*Given up:* a user can under-evaluate by habit — narrow to C2ST once, keep the call around, and never see Energy Distance when it lands, because an explicit list silently excludes every future metric. The mitigation is that `None` is the default and the docstring says so; there is no way to be additive-by-default and selectable at once.

---

### New vocabulary

`evaluate_imputation`, `EvaluationReport`, `EvaluationMetric`, and the rule **observability follows loop ownership**. Destined for the `CONTEXT.md` distillation at the map's close.

### Amendments this forces on `CONTEXT.md` (at the map's close, not now)

- **Observability** — the Progress Observer is supplied *per orchestrator call*, not *per orchestrator constructor*; **Phase Orchestrator** no longer implies a class.
- **Observability** — state the loop-ownership rule explicitly, so the `fit_unit`-emits-nothing / evaluation-emits-Events split reads as a rule rather than an inconsistency.
- **Pipeline Event** — `phase` is a plain string and is not always a `PipelinePhase` member; `"evaluation"` is deliberately absent from the enum.
- **Public API** — `EvaluationOrchestrator` is removed; `evaluate_imputation`, `EvaluationReport` and `EvaluationMetric` are added.

### Inputs handed to [Deletion inventory for the old evaluation surface](https://github.com/DEVunderdog/DataForgeML/issues/492)

Not decided here, but fixed by this ticket and binding on that one:

- `EvaluationOrchestrator` the **class** goes, not merely its methods.
- Evaluation's `ThreadPoolExecutor` and `_resolve_fit_workers` go, and no `max_workers` dial appears anywhere on `EvaluationConfig`.
- `_ObservabilityMixin` **stays** — the profiler still uses it.

**Comment by DEVunderdog** [comment]
Resolved — see the resolution comment above. Events stay because the library owns the loop; no pool anywhere because HGB is already threaded; `evaluate_imputation` is a free function taking `observer=` and `metrics=` and returning an `EvaluationReport` envelope.
